python/tk_multi_workfiles/entity_models/deferred_model.py

```diff
@@ -0,0 +1,588 @@
+                continue
+            if self.canFetchMore(item.index()):
+                self.fetchMore(item.index())
+            for row_i in range(item.rowCount()):
+                child_item = item.child(row_i)
+                item_list.append(child_item)
+
+    def clear(self):
+        """
+        Clear the data we hold.
+        """
+        super(ShotgunDeferredEntityModel, self).clear()
+        # Get a fresh and empty deferred cache.
+        self._deferred_cache = ShotgunDataHandlerCache()
+        # TODO: clear deferred models?
+
+    def hasChildren(self, index):
+        """
+        Return True if the item at the given index has children.
+
+        :param index: A :class:`QtCore.QModelIndex` instance.
+        :returns: A boolean, whether or not the given index has children.
+        """
+        # Just call base implementation if the index is not valid or if we are
+        # not dealing with a leaf.
+        if not index.isValid() or not self.itemFromIndex(index).get_sg_data():
+            return super(ShotgunDeferredEntityModel, self).hasChildren(index)
+        # We always have at least a child, which can be a valid item pulled with
+        # a deferred query or a place holder item.
+        return True
+
+    def canFetchMore(self, index):
+        """
+        Return True if more children can be fetched under the given index.
+
+        :param index: A :class:`QtCore.QModelIndex` instance.
+        :returns: A boolean, whether or not more children can be fetched.
+        """
+        # Just call base implementation if the index is not valid or if we are
+        # not dealing with a leaf.
+        if not index.isValid() or not self.itemFromIndex(index).get_sg_data():
+            return super(ShotgunDeferredEntityModel, self).canFetchMore(index)
+
+        item = self.itemFromIndex(index)
+        if item.data(self._SG_ITEM_FETCHED_MORE):
+            # More data has already been queried for this item
+            return False
+        return True
+
+    def fetchMore(self, index):
+        """
+        Fetch more children under the given index.
+
+        :param index: A :class:`QtCore.QModelIndex` instance.
+        """
+        if not index.isValid():
+            # Let the base implementation deal with invalid items
+            return super(ShotgunDeferredEntityModel, self).fetchMore(index)
+        item = self.itemFromIndex(index)
+        # Set the flag to prevent subsequent attempts to fetch more
+        item.setData(True, self._SG_ITEM_FETCHED_MORE)
+        sg_data = item.get_sg_data()
+        if not sg_data:
+            # If not dealing with a leaf, let the base implementation deal with
+            # the index.
+            return super(ShotgunDeferredEntityModel, self).fetchMore(index)
+        sub_entities = self._run_deferred_query_for_entity(sg_data)
+
+    def item_from_entity(self, entity_type, entity_id):
+        """
+        Retrieve the item representing the given entity in the model.
```

Comment:
what model is this? only the 'main' one? All loaded child models? How does the deferred-ness play into what is being returned?

Reply:
From the outside world there is a single model. Adding a comment about the deferredness.